### PR TITLE
Add testing provisions

### PR DESCRIPTION
**Pull request by bifurcation**
This PR adds two documents:

* A proposal for an interop testing framework
* A proposed initial set of scenarios to test

### CODE REVIEW

File: test-harness.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  * Inputs: PublicGroupState, encrypt flag
+  * Outputs: MLSPlaintext(Commit), Welcome, pointer to new state
+
+Operations using a group state:
+
+* Get state auth
+  * Inputs: (none)
+  * Outputs: authentication_secret
+* Export a secret
+  * Inputs: label, context
```
Comment (raphaelrobert) [inline test-harness.md:39]: Is context the secret length?

File: test-harness.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+  * Outputs: MLSPlaintext(Proposal(ReInit))
+* Generate an AppAck proposal for all messages processed in this epoch
+  * Inputs: (none)
+  * Outputs: MLSPlaintext(Proposal(AppAck))
+* Generate a Commit message covering a set of proposals
+  * Inputs: One or more MLSPlaintext(Proposal(\*))
+  * Outputs: Zero or more MLSPlaintext(Proposal(\*)), MLSPlaintext(Commit)
+  * Note: The committer chooses whether to inline any proposals, and returns the
+    proposals not inlined.  If they commiter can't commit one or more of the
+    proposals, it must return an error.
```
Comment (raphaelrobert) [inline test-harness.md:75]: I think we should be more flexible here. The input proposals don't necessarily have to be from the the committer, they could also be from another sender. In addition, I think the inline proposals should be a separate input.

That way it would cover real-world scenarios where you could have a complete mix of proposals that are covered by a Commit: (own stand-alone proposals, stand-alone proposals from others and inline proposals)

File: test-harness.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+* Handle a collection of MLSPlaintexts and a Commit
+  * Inputs: Zero or more MLSPlaintext(Proposal(\*)), MLSPlaintext(Commit)
+  * Outputs: pointer to new state
```
Comment (raphaelrobert) [inline test-harness.md:79]: The output should cover errors

File: test-harness.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+* Manually specified HTTP API
+* HTTP API structured with something like [RAML](https://raml.org/) or
+  [Swagger](https://swagger.io/)
+* Using an RPC framework such as [gRPC](https://grpc.io/) or
+  [Thrift](https://thrift.apache.org/)
+
+All of these options allow implementations in multiple languages and on multiple
+platforms.  Using a pre-provided RPC framework could simplify some RPC details,
+but would require implementors to figure out how ti integrate that framework.
```
Comment (raphaelrobert) [inline test-harness.md:94]: "to integrate"

File: test-scenarios.md [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+## Verifying Common State (subroutine)
+
+Since the major goal of MLS is to synchronize cryptographic state across a
+group's members, the key factor for verifying interop is verifying whether
+clients do indeed have the same cryptographic state.  The following things
+should be checked:
+
+* Each member's state produces the same authentication secret
+* An export with the same label and context produces the same exported value at
+  each member
```
Comment (raphaelrobert) [inline test-scenarios.md:21]: I think the export secret and the authentication secret are somewhat redundant. It would be enough to test them once and then only use one of them here.

File: test-scenarios.md [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+B->A: Welcome(new group ID, reinit PSK)
+***:  Verify group state
+```
+
+## Large Group, Full Lifecycle
+
+* Create group
+* Group creator adds the first M members
+* Until group size reaches N members, a randomly-chosen group member adds a new
+  member
```
Comment (raphaelrobert) [inline test-scenarios.md:135]: I think we want to specify that we should only use full Commits when adding members, so that the tree get more populated and we cover more edge-cases.

File: test-scenarios.md [hunk 7]
```diff
@@ -0,0 +1,139 @@
+MLS Interop Test Scenarios
+==========================
```
Comment (raphaelrobert) [inline test-scenarios.md:2]: So these are essentially scripts the test runner would execute among the different implementations?

File: test-harness.md [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+  * Inputs: ciphersuite, credential, signature private key
+  * Outputs: KeyPackage, callback to which Welcome should be sent
+* Complete joining a group (Welcome callback)
+  * Inputs: Welcome, encrypt flag
+  * Outputs: Pointer to new state
+* External join
+  * Inputs: PublicGroupState, encrypt flag
+  * Outputs: MLSPlaintext(Commit), Welcome, pointer to new state
+
+Operations using a group state:
```
Comment (raphaelrobert) [inline test-harness.md:33]: Some of these operations (probably most of them) can also produce errors. In a second step it would be good to test implementations with faulty input data.

**Review by raphaelrobert — CHANGES_REQUESTED** [review]
Good start! I made some comments here and there.

File: test-harness.md (hunk 1 above)
Comment (bifurcation) [reply test-harness.md:39]: No, but we do need a length parameter.  The MLS exporter [takes three arguments](https://github.com/mlswg/mls-protocol/blob/master/draft-ietf-mls-protocol.md#exporters), label, context, and length.

File: test-harness.md (hunk 2 above)
Comment (bifurcation) [reply test-harness.md:75]: Good idea.  I'll update to allow the caller to specify inline / reference proposals separately.

File: test-harness.md (hunk 3 above)
Comment (bifurcation) [reply test-harness.md:79]: I think this is true in general; everything should be fallible.  Adding a note to the top.

File: test-scenarios.md (hunk 5 above)
Comment (bifurcation) [reply test-scenarios.md:21]: Yeah, this seems sensible.  We can validate the correctness of the export logic in test vectors (which I have added to the test harness interface).

File: test-scenarios.md (hunk 6 above)
Comment (bifurcation) [reply test-scenarios.md:135]: Well, as long as we only have `A` doing adds, we'll only populate that one path.  But we should cover both full and non-full adds.

File: test-scenarios.md (hunk 7 above)
Comment (bifurcation) [reply test-scenarios.md:2]: Yep, that's the idea.

File: test-harness.md (hunk 8 above)
Comment (bifurcation) [reply test-harness.md:33]: I've marked everything as fallible, so we should be able to produce errors.  I think since we're passing in bytes here, we should also be able to pass in malformed bytes.  The main question is what malformed data we should provide and at what points, which I think is a good question for the "scenarios"
